reanalysis/validate_categories.py

```diff
@@ -0,0 +1,344 @@
+            # -
+            # if the gene isn't 'new' in PanelApp, remove Class2 flag
+            # in future expand to the 'if MOI has changed' logic
+            if variant.category_2 and not panel_gene_data.get('new', False):
+                variant.category_2 = False
+
+            # we never use a C4-only variant as a principal variant
+            # and we don't consider a variant with no assigned classes
+            #   - no classes shouldn't have reached this point
+            if variant.category_4_only or not variant.is_classified:
+                continue
+
+            # we've decided to run MOI tests on this variant
+            # - find the simplified version of the MOI string
+            # - use it to get the appropriate MOI model
+            # - run the variant, and append relevant classification(s) to the results
+            results.extend(
+                moi_lookup[get_simple_moi(panelapp_data[gene].get('moi'))].run(
+                    principal_var=variant, ensg=gene
+                )
+            )
+
+    return results
+
+
+def clean_initial_results(
+    result_list: List[ReportedVariant],
+) -> Dict[str, Dict[str, ReportedVariant]]:
+    """
+    Possibility 1 variant can be classified multiple ways
+    This cleans those to unique for final report
+    Join all possible classes for the condensed variants
+    :param result_list:
+    """
+
+    clean_results: Dict[str, Dict[str, ReportedVariant]] = {}
+
+    for each_instance in result_list:
+        support_id = (
+            ','.join(sorted(each_instance.support_vars))
+            if each_instance.support_vars is not None
+            else 'Unsupported'
+        )
+        var_uid = (
+            f'{each_instance.var_data.coords.string_format}__'
+            f'{each_instance.gene}__'
+            f'{support_id}'
+        )
+
+        # create a section for this sample if it doesn't exist
+        if each_instance.sample not in clean_results:
+            clean_results[each_instance.sample] = {}
+
+        # get an existing object, or use the current one
+        variant_object = clean_results.setdefault(each_instance.sample, {}).setdefault(
+            var_uid, each_instance
+        )
+
+        # combine any possible reasons, and add
+        clean_results[each_instance.sample][
+            var_uid
+        ].reasons = variant_object.reasons.union(each_instance.reasons)
+    return clean_results
+
+
+def main(
+    labelled_vcf: str,
+    comp_het: str,
+    config_path: Union[str, Dict[str, Any]],
+    out_json: str,
+    panelapp: str,
+    pedigree: str,
+):
+    """
+    All VCFs in use at this point will be small
+    These have been pre-filtered to retain only a small number of classified variants
+    holding all the variants in memory should not be a challenge, no matter how large
+    the cohort; if the variant number is large, the classes should be refined
+    We expect approximately linear scaling with participants in the joint call
```

Comment:
This is a key design decision that might be worth pulling out to the docs and also worth verifying on real data :).

Reply:
I've tried to clarify my expectation here, i.e. the small memory footprint of each individual variant, and the intention that the overall variant counts be v. low (e.g. 1 per variant at the end of the process). This should be a self-fulfilling scaling... If this stage of the process contains too many variants, the solution will be to revisit the filtering stage to increase the signal-noise ratio by increasing stringency.

It is a little vague still, and remains to see more real-world usage on different callsets